Gems/SceneProcessing/Code/Tests/MeshBuilder/MeshOptimizerComponentTests.cpp

```diff
@@ -0,0 +1,221 @@
+            skinWeights->GetBoneId("0");
+            skinWeights->GetBoneId("1");
+
+            skinWeights->AppendLink(0, {/*.boneId=*/0, /*.weight=*/1});
+            skinWeights->AppendLink(1, {/*.boneId=*/0, /*.weight=*/1});
+            skinWeights->AppendLink(2, {/*.boneId=*/0, /*.weight=*/1});
+            skinWeights->AppendLink(3, {/*.boneId=*/1, /*.weight=*/1});
+            skinWeights->AppendLink(4, {/*.boneId=*/1, /*.weight=*/1});
+            skinWeights->AppendLink(5, {/*.boneId=*/1, /*.weight=*/1});
+
+            return skinWeights;
+        }
+
+    private:
+        AZ::ComponentApplication m_app;
+        AZ::Entity* m_systemEntity;
+    };
+
+    TEST_F(VertexDeduplicationFixture, CanDeduplicateVertices)
+    {
+        AZ::SceneAPI::Containers::Scene scene("testScene");
+        AZ::SceneAPI::Containers::SceneGraph& graph = scene.GetGraph();
+
+        const auto meshNodeIndex = graph.AddChild(graph.GetRoot(), "testMesh", MakePlaneMesh());
+
+        // The original source mesh should have 6 vertices
+        EXPECT_EQ(AZStd::rtti_pointer_cast<AZ::SceneAPI::DataTypes::IMeshData>(graph.GetNodeContent(meshNodeIndex))->GetVertexCount(), 6);
+
+        auto meshGroup = AZStd::make_unique<AZ::SceneAPI::SceneData::MeshGroup>();
+        meshGroup->GetSceneNodeSelectionList().AddSelectedNode("testMesh");
+        scene.GetManifest().AddEntry(AZStd::move(meshGroup));
+
+        AZ::SceneGenerationComponents::MeshOptimizerComponent component;
+        AZ::SceneAPI::Events::GenerateSimplificationEventContext context(scene, "pc");
+        component.OptimizeMeshes(context);
+
+        AZ::SceneAPI::Containers::SceneGraph::NodeIndex optimizedNodeIndex = graph.Find(AZStd::string("testMesh").append(AZ::SceneAPI::Utilities::OptimizedMeshSuffix));
+        ASSERT_TRUE(optimizedNodeIndex.IsValid()) << "Mesh optimizer did not add an optimized version of the mesh";
+
+        const auto& optimizedMesh = AZStd::rtti_pointer_cast<AZ::SceneAPI::DataTypes::IMeshData>(graph.GetNodeContent(optimizedNodeIndex));
+        ASSERT_TRUE(optimizedMesh);
+
+        // The optimized mesh should have 4 vertices, the 2 shared vertices are welded together
+        EXPECT_EQ(optimizedMesh->GetVertexCount(), 4);
+    }
+
+    MATCHER(VectorOfLinksEq, "")
+    {
+        return testing::ExplainMatchResult(
+            testing::AllOf(
+                testing::Field(&AZ::SceneData::GraphData::SkinWeightData::Link::boneId, testing::Eq(testing::get<0>(arg).boneId)),
+                testing::Field(&AZ::SceneData::GraphData::SkinWeightData::Link::weight, testing::FloatEq(testing::get<0>(arg).weight))
+            ),
+            testing::get<1>(arg),
+            result_listener
+        );
+    }
+
+    MATCHER(VectorOfVectorOfLinksEq, "")
+    {
+        return testing::ExplainMatchResult(
+            testing::UnorderedPointwise(VectorOfLinksEq(), testing::get<0>(arg)),
+            testing::get<1>(arg),
+            result_listener
+        );
+    }
+
```

Comment:
Nitpick: I think some additional tests would be helpful here. Edge cases, things like that:

- Test a mesh where none of the verts are identical, make sure nothing gets removed.
- Test with a mesh that has no vertices or faces.
- Test a mesh where all vertices are identical.
- Test a mesh that's not properly setup, it has vertices but no faces, and another with faces and no vertices.
- Maybe a performance test, but those are a challenge with the wide range of hardware that these tests can be run on.

Reply:
I think it makes sense to test meshes that could actually exist in the scene graph, but I believe the scene graph already avoids adding invalid meshes to the graph. Since this component consumes a scene graph, and not a raw mesh file that could have those edge cases, do you think it is worth it to add tests for those?

Reply:
I could go either way there. The main benefit of testing things that are normally handled up or downstream like that, is it makes this code more robust if that behavior ever changes. It's unlikely anyone will go and strip the code that handles invalid meshes in the scene graph, but it isn't unusual that someone might change exactly how that's handled.

As an example, there could be some new product output of the Scene Builder that consumes meshes that were previously considered invalid, so the scene graph is updated to support those meshes going through.